Src/ILGPU/Context.cs

```diff
@@ -384,6 +385,77 @@ public DeviceCollection<TDevice> GetDevices<TDevice>()
                 : new DeviceCollection<TDevice>(new List<Device>());
         }
 
+        /// <summary>
+        /// Attempts to return the most optimal single device.
+        /// </summary>
+        /// <param name="PreferCPU">Always returns CPU device 0.</param>
+        /// <returns>Selected device.</returns>
+        public Device GetBestDevice(bool PreferCPU = false)
+        {
+            if(PreferCPU)
+            {
+                return deviceMapping.TryGetValue(AcceleratorType.CPU, out var devices)
+                    ? devices.First()
+                    : throw new NotSupportedException(
+                            RuntimeErrorMessages.NotSupportedTargetAccelerator);
+            }
+
+            var sorted = Devices.Sort((d1, d2) => d1.MemorySize.CompareTo(d2.MemorySize))
+                .Where(d => d.AcceleratorType != AcceleratorType.CPU);
+
+            if(sorted.Count() < 0)
```

Comment:
ISSUE: When the `sorted.Count() == 0`, we are going to call `sorted.First()` which will throw an exception.

Also, not optimal to call `Count()` if you don't need the number.
Maybe just `return sorted.FirstOrDefault() ?? GetCPUDevice(0);`

Reply:
I replaced sorted.Count with sorted,Any() in the GetPreferredDevices, in GetPreferredDevice I used the return sorted.FirstOrDefault() ?? GetCPUDevice(0);